### PR TITLE
Infra: Shard spark and spark master tests across 3 actions

### PR DESCRIPTION
**Pull request by amogh-jahagirdar**
This change adds a matrix of 3 shards to github CI for the delta spark/spark master tests and plumbs those through to our parallelization strategy which will then select tests to run if they are assigned to that shard. This brings down CI times from ~4 hours to 2 hours.

3 shards seemsd to be the point of diminishing returns, and we should look at the actual tests and possibly higher instance types if we really want to get those down further.

No sharding: 4 hours
2 shards: 3 hours
3 shards:  ~2 hours
4 shards: ~2 hours

### How to revert

To undo this change without a full revert commit there are a few ways (in order of recommendation)

1. Remove/comment out *both* 
the shard row in the GH matrix https://github.com/delta-io/delta/pull/3517/files#diff-b77047dc65c62d814f7f67eec57c23c4cf9d9796e6a8a018558e0e935739d8d8R10-R11

AND

the NUM_SHARDS env variable
https://github.com/delta-io/delta/pull/3517/files#diff-b77047dc65c62d814f7f67eec57c23c4cf9d9796e6a8a018558e0e935739d8d8R15

2. Comment out this block https://github.com/delta-io/delta/pull/3517/files#diff-00b77d52981e1ba8cc5ac9b80b22d8f7bb9814222121b714269e14e139daaae0R141-R152 in TestParallelization

### CODE REVIEW

**Comment by amogh-jahagirdar** [comment]
cc @nastra @scottsand-db @tdas

File: build.sbt [hunk 1]
```diff
@@ -1506,8 +1506,7 @@ val createTargetClassesDir = taskKey[Unit]("create target classes dir")
 
 // Don't use these groups for any other projects
 lazy val sparkGroup = project
-  .aggregate(spark, contribs, storage, storageS3DynamoDB, iceberg, testDeltaIcebergJar, sharing, hudi)
-  .settings(
+  .aggregate(spark, contribs, storage, storageS3DynamoDB, iceberg, testDeltaIcebergJar, sharing, hudi)  .settings(
```
Comment (nastra) [inline build.sbt:1509]: nit: unnecessary change

File: .github/workflows/spark_master_test.yaml [hunk 2]
```diff
@@ -46,7 +48,7 @@ jobs:
       - name: Run Spark Master tests
         # when changing TEST_PARALLELISM_COUNT make sure to also change it in spark_test.yaml
         run: |
-          TEST_PARALLELISM_COUNT=2 build/sbt -DsparkVersion=master "++ ${{ matrix.scala }}" clean spark/test
+          TEST_PARALLELISM_COUNT=2 SHARD_ID=${{matrix.shard}} build/sbt -DsparkVersion=master "++ ${{ matrix.scala }}" clean spark/test
```
Comment (scottsand-db) [inline .github/workflows/spark_master_test.yaml:51]: how does this select just a single shard integer? To me this looks like it would select the entire array

File: build.sbt (hunk 1 above)
Comment (scottsand-db) [reply build.sbt:1509]: +1

File: run-tests.py [hunk 3]
```diff
@@ -43,16 +43,24 @@ def get_args():
         default=False,
         action="store_true",
         help="Enables test coverage and generates an aggregate report for all subprojects")
+
```
Comment (scottsand-db) [inline run-tests.py:46]: nit: delete white space

File: project/TestParallelization.scala [hunk 4]
```diff
@@ -50,17 +52,24 @@ object TestParallelization {
     Test / forkTestJVMCount := {
       sys.env.get("TEST_PARALLELISM_COUNT").map(_.toInt).getOrElse(4)
     },
+    Test / shardId := {
+      sys.env.get("SHARD_ID").map(_.toInt)
```
Comment (scottsand-db) [inline project/TestParallelization.scala:56]: can you use `numShards` here?

File: .github/workflows/spark_master_test.yaml (hunk 2 above)
Comment (amogh-jahagirdar) [reply .github/workflows/spark_master_test.yaml:51]: The GH matrix option will tell GH actions to trigger a job per matrix combination. https://docs.github.com/en/actions/writing-workflows/choosing-what-your-workflow-does/using-a-matrix-for-your-jobs

So in this case for spark master there will be a  job for Scala 2.13.13 for shard 0, another one for 1, for 2 etc.

matrix.{value} always references what is the concrete value for that given job. 
For example, It's how the delta spark tests work for testing the different scala versions (2.12.8 and 2.13.13 in spark_test.yaml). Those scala versions ultimately get passed through to `run-tests.py`

File: build.sbt (hunk 1 above)
Comment (amogh-jahagirdar) [reply build.sbt:1509]: Sure will fix this

File: project/TestParallelization.scala (hunk 4 above)
Comment (amogh-jahagirdar) [reply project/TestParallelization.scala:56]: No, that wouldn't be correct, this is supposed to be the shard ID assigned to a given action.
`numShards` is just used for the modulo math to determine if a test is assigned to the given shard ID or not. I should probably comment this, or if there's better names, open to suggestion.

**Comment by amogh-jahagirdar** [comment]
`CloneIcebergSuiteByName` is failing after I rebased, I'm retriggering to see if it's flakey

File: project/TestParallelization.scala (hunk 4 above)
Comment (scottsand-db) [reply project/TestParallelization.scala:56]: Oh, sorry, I just misread this. Ignore this comment. Thanks!

File: project/TestParallelization.scala [hunk 5]
```diff
@@ -107,25 +116,38 @@ object TestParallelization {
...
+  class SimpleHashStrategy private(
+      groups: Map[Int, Tests.Group],
+      shard: Option[Int]) extends GroupingStrategy {
 
     lazy val testGroups = groups.values.toList
     val groupCount = groups.size
 
     override def add(testDefinition: TestDefinition): GroupingStrategy = {
       val groupIdx = math.abs(testDefinition.name.hashCode % groupCount)
+      if (shard.isDefined) {
```
Comment (scottsand-db) [inline project/TestParallelization.scala:128]: add some comments here to explain this please. After a few minutes I still don't 100% understand the logic

Comment (scottsand-db) [inline project/TestParallelization.scala:128]: if we wanted to disable this, say 6 months from now due to some crazy bug, the minimal change we would have to do is just disable this if statement, right? e.g. `if (false)`, right?

Comment (amogh-jahagirdar) [reply project/TestParallelization.scala]: I updated with a comment, please let me know if it still is unclear, I think that's important to make sure everyone's confident on this sharding logic!

Comment (amogh-jahagirdar) [reply project/TestParallelization.scala]: Correct, that's def a way to do it in the code itself. From the GH configuration perspective, it's possible to also just comment out the 2 lines for matrix shard ID list and the shard ID env variable.

File: .github/workflows/spark_test.yaml [hunk 6]
```diff
@@ -7,8 +7,10 @@ jobs:
       matrix:
         # These Scala versions must match those in the build.sbt
         scala: [2.12.18, 2.13.13]
+        shard: [0, 1, 2]
     env:
       SCALA_VERSION: ${{ matrix.scala }}
+      NUM_SHARDS: 3
```
[the remark this answers is not in the document]
Comment (amogh-jahagirdar) [reply .github/workflows/spark_test.yaml:13]: https://github.com/delta-io/delta/pull/3517/files#diff-b77047dc65c62d814f7f67eec57c23c4cf9d9796e6a8a018558e0e935739d8d8R10 

The Github matrix strategies itself. Github CI itself will read the matrix fields, and create 1 job per matrix combination. It's important NUM_SHARDS == length (matrix.shards) I've commented that in the Yaml so anyone who tries to change this will be super aware of the assumptions, and I also commented how the sharding logic works in TestParallelization.

We just need the environment variable NUM_SHARDS so that in TestParallelization we can determine if a given test is assigned to a shard based on % NUM_SHARDS == the assigned shard ID.

https://docs.github.com/en/actions/writing-workflows/choosing-what-your-workflow-does/using-a-matrix-for-your-jobs#about-matrix-strategies

File: project/TestParallelization.scala (hunk 5 above)
Comment (scottsand-db) [reply project/TestParallelization.scala]: @amogh-jahagirdar - please include this "how to disable later" in the PR description. That will make it super findable.

**Review by scottsand-db — APPROVED** [review]
